packages/@pollyjs/core/src/-private/event-emitter.js

```diff
@@ -0,0 +1,218 @@
+import { assert } from '@pollyjs/utils';
+
+const EVENTS = Symbol();
+const EVENT_NAMES = Symbol();
+
+function assertEventName(eventName, eventNames) {
+  assert(
+    `Invalid event name provided. Expected string, received: "${typeof eventName}".`,
+    typeof eventName === 'string'
+  );
+
+  assert(
+    `Invalid event name provided: "${eventName}". Possible events: ${[
+      ...eventNames
+    ].join(', ')}.`,
+    eventNames.has(eventName)
+  );
+}
+
+function assertListener(listener) {
+  assert(
+    `Invalid listener provided. Expected function, received: "${typeof listener}".`,
+    typeof listener === 'function'
+  );
+}
+
+export default class EventEmitter {
```

Comment:
Does anything event emitters out there implement a similar API?  https://github.com/sindresorhus/emittery is one that I know of that may

Reply:
Yeah I found that one too but it was lacking the ability to:

1. Get all the listeners for an event
2. Emit events synchronously 
3. Pass more than one argument when emitting an event (which would break the api)

Reply:
Fair points, lets continue with what you have.  One nit, I think we should rally around async vs supporting both.  Makes the API a bit confusing since depending on the caller, your handler may fire async or sync.

I see the one place that it is sync is with the create event.  This makes sense since you're not really able to await there.  But we can throw if the consumer is returning a promise there which can be detected at the end of the constructor if the promise is still not settled (Promise.race to detect that).

Reply:
I can actually detect that from the emitSync method. I can check if the return value is promise like and throw if it is.

Reply:
But emitSync sticks around and over time could get confusing.  "Is this an async event or sync event?"